File: C++/insertion_sort.cpp
#include <iostream>

void insert_sort(int array[], int tamanho){
    for(int i = 0; i < tamanho; i++){
        int valor_corrente = array[i];
        int j = i - 1;
        while((j >= 0) && (array[j] > valor_corrente)){
            array[j+1] = array[j];
            j--;
        }
        array[j+1] = valor_corrente;
    }
}

void busca(int array[], int tamanho, int valor){
    bool encontrado = false;

    for(int i = 0; i < tamanho; i++){
        if(valor == array[i]){
            if(valor == array[0]){
                std::cout << "O valor " << valor << " está na posição 0." << std::endl;
            }
            else{
                std::cout << "O valor está na posição " << i << ". Onde o valor " << array[0] << " se encontra na posição 0." << std::endl;
            }
        encontrado = true;
        }
    }
    if(encontrado == false){
        std::cout << "O valor " << valor << " não está no vetor." << std::endl;
    }
}

int main(){
    int valor, array[] = {30, 28, 83, 14, 56, 356, 97, 108, 9999, 200}, tamanho = sizeof(array)/sizeof(array[0]);

    insert_sort(array, tamanho);

    std::cout << "Escolha o número que deseja buscar no vetor [";
    for(int i = 0; i < tamanho; i++){
        std::cout << array[i];
        if(i < tamanho - 1){
            std::cout << ", ";
        }
    }
    std::cout << "]: ";
    std::cin >> valor;

    busca(array, tamanho, valor);

    return 0;
}

File: python/insertion_sort.py
def insertion_sort(array):
    for i in range(1, len(array)):
        valor_corrente = array[i]
        j = i - 1
        while j >= 0 and array[j] > valor_corrente:
            array[j+1] = array[j]
            j-= 1
        array[j+1] = valor_corrente
    return array

def busca(array, numero):
    encontrado = False
    for i in range(0, len(array)):
        if numero == array[i]:
            if numero == array[0]:
                print(f"O número {numero} está na posição 0")
            else:
                print(f"O número {numero} está presente na posição {i}. Onde o {array[0]} se encontra na posição 0.")
            encontrado = True
    if encontrado == False:
        print(f"O numero {numero} não está presente no array.")
    

array = [5, 1, 39, 28, 104, 90, 80, -39]

insertion_sort(array)

numero = int(input(f"Digite o número que deseja encontrar no  array {array}: "))

busca(array, numero)